addons/mail/models/res_users.py

```diff
@@ -115,45 +117,43 @@ def _unsubscribe_from_channels(self):
 
     @api.model
     def systray_get_activities(self):
-        query = """SELECT m.id, count(*), act.res_model as model,
-                        CASE
-                            WHEN %(today)s::date - act.date_deadline::date = 0 Then 'today'
-                            WHEN %(today)s::date - act.date_deadline::date > 0 Then 'overdue'
-                            WHEN %(today)s::date - act.date_deadline::date < 0 Then 'planned'
-                        END AS states
-                    FROM mail_activity AS act
-                    JOIN ir_model AS m ON act.res_model_id = m.id
-                    WHERE user_id = %(user_id)s
-                    GROUP BY m.id, states, act.res_model;
-                    """
-        self.env.cr.execute(query, {
-            'today': fields.Date.context_today(self),
-            'user_id': self.env.uid,
-        })
-        activity_data = self.env.cr.dictfetchall()
-        model_ids = [a['id'] for a in activity_data]
-        model_names = {n[0]: n[1] for n in self.env['ir.model'].browse(model_ids).name_get()}
-
+        activities = self.env["mail.activity"].search([("user_id", "=", self.env.uid)])
+        activities_by_record_by_model_name = defaultdict(lambda: defaultdict(lambda: self.env["mail.activity"]))
+        for activity in activities:
+            record = self.env[activity.res_model].browse(activity.res_id)
```

Comment:
note : you don't need browse record because after you use only ids

`activities_by_record_by_model_name[activity.res_model][activity.res_id]`

Reply:
`browse` is lightweight, it's better to work with recordsets than with ids. Also changing this would mean naming the variable `activities_by_record_id_by_model_name` rather than `activities_by_record_by_model_name` which makes it a bit more complex everywhere else.